### ISSUE TITLE
windows: panic > runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by Dexus** [dgraph-io/badger#128]
```
 $ go run /tmp/badger.go
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x60 pc=0x6d2509]

goroutine 1 [running]:
github.com/dgraph-io/badger.(*KV).sendToWriteCh(0x0, 0xc042079d90, 0x1, 0x1, 0x0, 0x0, 0xc042079d50)
        C:/GoProjects/src/github.com/dgraph-io/badger/kv.go:699 +0xa9
github.com/dgraph-io/badger.(*KV).BatchSet(0x0, 0xc042079d90, 0x1, 0x1, 0x8, 0x5)
        C:/GoProjects/src/github.com/dgraph-io/badger/kv.go:723 +0x5d
github.com/dgraph-io/badger.(*KV).Set(0x0, 0xc042034fe0, 0x5, 0x8, 0xc042034fe8, 0x5, 0x8, 0x0, 0xc042079e88)
        C:/GoProjects/src/github.com/dgraph-io/badger/kv.go:763 +0xce
main.main()
        C:/tmp/badger.go:40 +0x1d1
exit status 2
```

```
package main

import (
	"fmt"
	"io/ioutil"

	"github.com/dgraph-io/badger"
)

var d string = "doc"

func main() {
	opt := badger.DefaultOptions
	dir, _ := ioutil.TempDir("/tmp", "badger")
	opt.Dir = dir
	opt.ValueDir = dir
	opt.LevelOneSize = 64 << 10
	opt.MaxTableSize = 32 << 10
	opt.ValueLogFileSize = 1 << 10
	kv, _ := badger.NewKV(&opt)

	key := []byte("hello")

	kv.Set(key, []byte("world"))
	fmt.Printf("SET %s world\n", key)

	var item badger.KVItem
	if err := kv.Get(key, &item); err != nil {
		fmt.Printf("Error while getting key: %q", key)
	}
	fmt.Printf("GET %s %s\n", key, item.Value())

	if err := kv.CompareAndSet(key, []byte("venus"), 100); err != nil {
		fmt.Println("CAS counter mismatch")
	} else {
		if err = kv.Get(key, &item); err != nil {
			fmt.Printf("Error while getting key: %q", key)
		}
		fmt.Printf("Set to %s\n", item.Value())
	}
	if err := kv.CompareAndSet(key, []byte("mars"), item.Counter()); err == nil {
		fmt.Println("Set to mars")
	} else {
		fmt.Printf("Unsuccessful write. Got error: %v\n", err)
	}
}
```

**Comment by Dexus** [comment]
It looks like #93 is not only on closed kv.

Maybe a better Test should be implemented...

**Comment by pawanrawal** [comment]
Please check for errors.

```
kv, err := badger.NewKV(&opt)
```

**Comment by Dexus** [comment]
```
Error: Directory entry for compaction log: sync C:\tmp\badger927045415: Das Handle ist ungültig.
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x60 pc=0x6d2509]

goroutine 1 [running]:
github.com/dgraph-io/badger.(*KV).sendToWriteCh(0x0, 0xc042025d80, 0x1, 0x1, 0x63, 0x0, 0xc042025d40)
        C:/GoProjects/src/github.com/dgraph-io/badger/kv.go:699 +0xa9
github.com/dgraph-io/badger.(*KV).BatchSet(0x0, 0xc042025d80, 0x1, 0x1, 0x8, 0x5)
        C:/GoProjects/src/github.com/dgraph-io/badger/kv.go:723 +0x5d
github.com/dgraph-io/badger.(*KV).Set(0x0, 0xc04200c510, 0x5, 0x8, 0xc04200c518, 0x5, 0x8, 0x0, 0x44fd50)
        C:/GoProjects/src/github.com/dgraph-io/badger/kv.go:763 +0xce
main.main()
        C:/tmp/badger.go:43 +0x22e
exit status 2
```

**Comment by pawanrawal** [comment]
```
dir, err := ioutil.TempDir("/tmp", "badger")
```
And here too, `/tmp` is not valid on Windows.

**Comment by Dexus** [comment]
It's doesn't matter if i us `/tmp` or an empty string there, the error is the same with the windows temp dir. 

Btw. Windows has with the developer mode (by ubuntu supported experimental [win10pro]) the `/tmp` dir on my pc, also with busy-box ;) so yes tmp dir is there and works like excepted.

Von meinem iPhone gesendet

**Comment by Dexus** [comment]
Mybe the use of the `os.PathSeparator` or the use of the `import "path/filepath"` in the right way would help? Only some thoughts... to fix the bug... when I'm up and out of the bed I'll create a fix PR.

**Comment by Motakjuq** [comment]
The crash is not related with the `os.PathSeparator`, it's a problem with kv.go file, specifically with `syncDir` function. It looks `f.Sync()` is not supported by windows. You have to work on an alternative way to perform the directory sync or remove the sync until someone fix `syncDir`.

**Comment by pawanrawal** [comment]
This is happening because we are trying to `fsync` a directory which Windows doesn't like and hence returns an error. I have to check if it's required on Windows.

Another small issue is the some our tests use `/tmp` while requesting a TempDir which doesn't work for Windows. I will send a PR to fix this.

**Comment by pawanrawal** [comment]
This issue should be fixed as of 39a53294ba445ab5b04278889042da6ef8e4b39b.
